core-patch/src/main/scala/eu/ostrzyciel/jelly/core/patch/internal/PatchDecoderImpl.scala

```diff
@@ -0,0 +1,188 @@
+      )
+
+    override protected def handleStatementDelete(statement: RdfQuad): Unit =
+      handler.deleteTriple(
+        convertTermWrapped(statement.subject, lastSubject),
+        convertTermWrapped(statement.predicate, lastPredicate),
+        convertTermWrapped(statement.`object`, lastObject),
+      )
+
+  private[core] final class QuadsDecoder[TNode, TDatatype : ClassTag](
+    converter: ProtoDecoderConverter[TNode, TDatatype, ?, ?],
+    handler: QuadPatchHandler[TNode],
+    supportedOptions: RdfPatchOptions,
+  ) extends PatchDecoderImpl[TNode, TDatatype](converter, handler, supportedOptions):
+
+    override protected def handleOptions(opt: RdfPatchOptions): Unit =
+      if !opt.statementType.isQuads then
+        throw new RdfProtoDeserializationError(
+          f"Incoming stream with statement type ${opt.statementType} cannot be decoded by this " +
+          f"decoder. Only QUADS streams are accepted.")
+      super.handleOptions(opt)
+
+    override protected def handleStatementAdd(statement: RdfQuad): Unit =
+      handler.addQuad(
+        convertTermWrapped(statement.subject, lastSubject),
+        convertTermWrapped(statement.predicate, lastPredicate),
+        convertTermWrapped(statement.`object`, lastObject),
+        convertGraphTermWrapped(statement.graph),
+      )
+
+    override protected def handleStatementDelete(statement: RdfQuad): Unit =
+      handler.deleteQuad(
+        convertTermWrapped(statement.subject, lastSubject),
+        convertTermWrapped(statement.predicate, lastPredicate),
+        convertTermWrapped(statement.`object`, lastObject),
+        convertGraphTermWrapped(statement.graph),
+      )
+
+  private[core] final class AnyStatementDecoder[TNode, TDatatype : ClassTag](
+    converter: ProtoDecoderConverter[TNode, TDatatype, ?, ?],
+    handler: AnyPatchHandler[TNode],
+    supportedOptions: RdfPatchOptions,
+  ) extends PatchDecoderImpl[TNode, TDatatype](converter, handler, supportedOptions):
+    private var inner: PatchDecoderImpl[TNode, TDatatype] = uninitialized
+
+    override def getPatchOpt: Option[RdfPatchOptions] =
+      if inner == null then None else inner.getPatchOpt
+
+    override def ingestRow(row: RdfPatchRow): Unit =
+      if inner == null then
+        if row.rowType != RdfPatchRow.OPTIONS_FIELD_NUMBER then
+          throw new RdfProtoDeserializationError(
+            "The first row in the stream must be an RdfPatchOptions row.")
+        else
+          val opt = row.row.asInstanceOf[RdfPatchOptions]
+          // We must call this in both this instance and the child instance to make sure
+          // both know how to handle punctuations.
+          handleOptions(opt)
+          createInnerDecoder(opt)
+          inner.ingestRow(row)
+      else
+        inner.ingestRow(row)
+
+    // Should not be called.
+    // We made these methods abstract to avoid the overhead of virtual function calls.
+    override def handleStatementAdd(statement: RdfQuad): Unit = ()
```

Comment:
I guess that it's because we move ingestRow logic to inner but why this specific pattern here